### ISSUE TITLE
h2o is MASSIVELY INSECURE BY DEFAULT, allows RC4, DES, 3DES in default config

### ISSUE THREAD

**Issue by DtpEJsaYXDU4GDH8dE4MyI9VrieF0UZpPZ0K76K** [h2o/h2o#1331]
h2o is MASSIVELY INSECURE BY DEFAULT, allows RC4, DES, 3DES in default config

these should not be enabled in any default config. it is not 1995 any more

**Comment by INRIX-joel-winarske** [comment]
Something like this is required:

```
#define SECURE_CIPHER_LIST "ECDH+AESGCM:DH+AESGCM:ECDH+AES256:DH+AES256:ECDH+AES128:DH+AES:RSA+AESGCM:RSA+AES:!aNULL:!MD5:!DSS"

if (SSL_CTX_set_cipher_list(accept_ctx.ssl_ctx, SECURE_CIPHER_LIST) != 1) {
    fprintf(stderr, "an error occurred while trying to set cipher list\n");
    return -1;
}
```
[hardening web server ssl ciphers](https://hynek.me/articles/hardening-your-web-servers-ssl-ciphers/)

**Comment by DtpEJsaYXDU4GDH8dE4MyI9VrieF0UZpPZ0K76K** [comment]
if that's a code change that enforces a minimum level of security, that's a good start.

but i think a much more secure default could be ensured with something like the following: 

minimum-version: TLSv1.2

cipher-suite: ECDHE-RSA-CHACHA20-POLY1305:ECDHE-RSA-AES256-GCM-SHA384:DHE-RSA-AES256-GCM-SHA384:ECDHE-RSA-AES128-GCM-SHA256;

cipher-preference: server

There is little reason to support anything less than TLSv1.2 and those robust cipher suites since all modern web clients support them now. the only ones that don't, to my knowledge, are end-of-life. thoughts?

**Comment by DtpEJsaYXDU4GDH8dE4MyI9VrieF0UZpPZ0K76K** [comment]
as an example of how poor and insecure h2o is by default, look at the ratings for h2o's own site (hint: "F" rating). pretty unsatisfactory if you ask me, but let's get it sorted and in a better position so that the future of the web runs on h2o and not nginx or apache. h2o is great, but should be secure out of the box if the project cares about user and organizational trust.

https://www.ssllabs.com/ssltest/analyze.html?d=h2o.examp1e.net

![1](https://user-images.githubusercontent.com/355620/27309918-992454ea-550b-11e7-935f-9cb6e0530d53.png)

**Comment by deweerdt** [comment]
Thanks for reporting this. I've started a PR so that we can discuss the changes there: https://github.com/h2o/h2o/pull/1334

**Comment by DtpEJsaYXDU4GDH8dE4MyI9VrieF0UZpPZ0K76K** [comment]
adoption of minimum version tls 1.2 is now happening. time to disable tls 1.0 and tls 1.1

https://lists.debian.org/debian-devel-announce/2017/08/msg00004.html

This will likely break certain things that for whatever reason
still don't support TLS 1.2. I strongly suggest that if it's not
supported that you add support for it, or get the other side to
add support for it.

OpenSSL made a release 5 years ago that supported TLS 1.2. The
current support of the server side seems to be around 90%.

**Comment by Varbin** [comment]
By the way: To exclude bad ciphers couldn't we just use something like `DEFAULT:!DES:!RC4:!MD5:!RC2`?

1. This is future proof: New ciphers in new Open/LibreSSL versions are automatically added,
2.  it deactivates insecure ciphers: Practically nobody needs DES (the 56 bit version, we may discuss about 3DES) anymore.
3. It uses the default list from Open/LibreSSL - it improves it.

**Comment by DtpEJsaYXDU4GDH8dE4MyI9VrieF0UZpPZ0K76K** [comment]
@Varbin yes, that's a start, but it's still not great. anything with "CBC" in it is exploitable by well funded adversaries and CBC are pretty popular cipher modes.

ideally, only strong forward secret cipher suites would be enabled by default, such as gcm, chacha20/poly1305, etc. forward secret ciphers help raise the cost of stolen tls key situations

**Comment by DtpEJsaYXDU4GDH8dE4MyI9VrieF0UZpPZ0K76K** [comment]
at the very least

https://wiki.mozilla.org/Security/Server_Side_TLS#Modern_compatibility

**Comment by concatime** [comment]
I think we should heavily invest in docs, and not really on sane defaults. In fact, if you want to use HTTPS, you're supposed to know that you minimally need some configurations, and it's not only magic! That's why I'm quite septic when I see how easy is CertBot.

**Comment by DtpEJsaYXDU4GDH8dE4MyI9VrieF0UZpPZ0K76K** [comment]
@concatime https configuration should be as simple out of the box as certbot

**Comment by concatime** [comment]
Simple, but not too much abstracted. HTTPS implies some responsibilities, that's a fact.

**Comment by Varbin** [comment]
@DtpEJsaYXDU4GDH8dE4MyI9VrieF0UZpPZ0K76K The idea is to find a balance - between absolute security (i.e. only allowing AES256 in GCM mode with ECDHE on TLS 1.2 ) and absolute usability including backwards compatibility (so also allowing end-of-life devices). If not doing so a piece of software will not be used by anyone. 
CBC issues can (as far I now) be mitigated (at least on client side) and so should be allowed as fallback. Modern devices would not loose anything -  they choose (or are forced to choose) modern cipher suites.

@kazuho I agree this all is an SSL library mistake - What do you think of my minimal idea? This will (hopefully) prevent total failure and connection issues (#1595) and is (partially) future proof.

`DEFAULT:!MD5:!DSS:!DES:!RC4:!RC2:!SEED:!IDEA:!NULL:!ADH:!EXP:!SRP:!PSK`

I tested this on multiple Open/LibreSSL versions and got no insecure ciphers and did not destroy backwards compatibility (so it should be safe to use).

**Comment by lkwg82** [comment]
@kazuho can you give a statement on this? So the next steps could be to offer PRs to improve the situation or to close this as inappropriate. Thx!

**Comment by DtpEJsaYXDU4GDH8dE4MyI9VrieF0UZpPZ0K76K** [comment]
@varbin an attacker can prevent a client from choosing the right cipher suites by a variety of techniques. So just keeping insecure suites around, by default, is still not a good idea. If someone wants them, let them enable a dangerous config flag manually that should scare anyone looking at the config considering to use insecure suites.

It was asked earlier that openssl should fix this upstream if this was really an issue. Well, that's exactly what they did shortly after such a discussion. A little while back openssl officially blacklisted older cipher suites and you have to specifically override them.

Allowing backward compatibility is fine, by being secure by default should take precedence. Let the admin make the choice of being insecure. Don't make it insecure for the people that may not even know how to edit a config.

Again, anything below TLS 1.2 should be killed. Any any cipher suite that is not forward secret should also be disallowed. The prior proposal config supports all currently supported clients with FS. Since it has been a while, perhaps it could even be updated a bit more since we have EdDSA, Ed25519, and others coming to replace RSA and ECDSA. At the present time, it would need to be tested how many supported clients actually can handle the new key exchange methods, but support is probably fairly low at the present time.
